kasa/klapprotocol.py

```diff
@@ -0,0 +1,515 @@
+"""Implementation of the TP-Link Smart Home Protocol.
+
+Comment by sdb - 4-Jul-2023
+
+Encryption/Decryption methods based on the works of
+Simon Wilkinson and Chris Weeldon
+
+While working on these changes I discovered my HS100 devices would periodically change their device owner to something that produces the following
+md5 owner hash: 994661e5222b8e5e3e1d90e73a322315.  It seems to be after an update to the on/off state that was scheduled via the app.  Switching the device on and off manually via the
+Kasa app would revert to the correct owner.
+
+For devices that have not been connected to the kasa cloud the theory is that blank username and password md5 hashes will succesfully authenticate but
+at this point I have been unable to verify.
+
+https://gist.github.com/chriswheeldon/3b17d974db3817613c69191c0480fe55
+https://github.com/python-kasa/python-kasa/pull/117
+
+N.B. chrisweeldon implementation had a bug in the encryption logic for determining the initial seq number and Simon Wilkinson's implementation did not seem to support
+incrementing the sequence number for subsequent encryption requests
+
+"""
+import asyncio
+import datetime
+import hashlib
+import logging
+import secrets
+from pprint import pformat as pf
+from typing import Any, Dict, Optional, Tuple, Union
+
+import aiohttp
+from cryptography.hazmat.primitives import hashes, padding
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from yarl import URL
```

Comment:
Could this (and other url construction mechanisms) be replaced with https://docs.python.org/3/library/urllib.parse.html#module-urllib.parse ?

Reply:
The cookie_jar update method expects this yarl class

Reply:
Yarl is definitely required by aiohttp.  For some reason the docs for AbstractCookieJar are wrong but if you follow the link to the source you see:
https://docs.aiohttp.org/en/stable/_modules/aiohttp/abc.html#AbstractCookieJar.update_cookies
and `poetry show yarl` output:
```
name         : yarl                    
 version      : 1.9.2                   
 description  : Yet another URL library 

dependencies
 - idna >=2.0
 - multidict >=4.0

required by
 - aiohttp >=1.0,<2.0
```

Reply:
No longer relevant since moving to httpx